STITCH-2194 assign content type based on file extension when no asset description for path exists

mpobrien opened the pull request:
had to move the asset description lookup into `FileToAssetMetadata`, otherwise the only way to make the tests pass would be to add the same code into `localFileToAssetMetadata` in the test code, but that would defeat the purpose of the test. Probably worth refactoring that test code at some point anyway.

mpobrien commented on commands/import.go:
```diff
@@ -228,7 +228,7 @@ func (ic *ImportCommand) importApp() error {
 		assetDescs, fileErr := hosting.MetadataFileToAssetDescriptions(filepath.Join(appPath, utils.HostingAttributes))
 		if fileErr != nil {
-			return errIncludeHosting(fileErr)
+			return errIncludeHosting(fmt.Errorf("error loading metadata.json file: %v", fileErr))
```
this is a drive-by to make it clearer to the user that the mistake exists specifically in the metadata.json file.

mburdette replied:
the `fileErr` is going to include the file name, so this might make the message stuttery

mpobrien replied:
It doesn't include the file name, it only has the JSON error. Without this change, it just looks like this:
`--include-hosting error: invalid character 's' looking for beginning of value`

mburdette commented on hosting/hosting_test.go:
```diff
@@ -30,7 +30,7 @@ func localFileToAssetMetadata(t *testing.T, localPath, rootDir string, assetDesc
 	filePath := fmt.Sprintf("/%s", relPath)
 	assetCache := hosting.NewAssetCache()
-	assetMetadata, famErr := hosting.FileToAssetMetadata(appID, localPath, filePath, info, assetDescriptions[filePath], assetCache)
+	assetMetadata, famErr := hosting.FileToAssetMetadata(appID, localPath, filePath, info, assetDescriptions, assetCache)
```
couldn't you just leave this as it was and pass the `assetDescriptions[filePath]` instead of having to pass all descriptions? This would then act as your `hasAssetDesc` bool and avoid passing all descriptions to every function call.

mburdette commented on hosting/hosting.go:
```diff
@@ -34,11 +34,7 @@ func buildAssetMetadata(appID string, assetMetadata *[]AssetMetadata, rootDir st
-				assetDesc = assetDescriptions[assetPath]
-			}
-			am, fileErr := FileToAssetMetadata(appID, path, assetPath, info, assetDesc, assetCache)
+			am, fileErr := FileToAssetMetadata(appID, path, assetPath, info, assetDescriptions, assetCache)
```
couldn't you just leave this as it was and pass the assetDescriptions[filePath] instead of having to pass all descriptions? This would then act as your hasAssetDesc bool and avoid passing all descriptions to every function call.

mpobrien replied:
I tried to do this originally, but then the tests break because the expected attributes aren't populated after calling `FileToAssetMetadata` - and to make them pass it would require updating the test function, `localFileToAssetMetadata` (https://github.com/10gen/stitch-cli/blob/master/hosting/hosting_test.go#L17-L43). But doing that would defeat the purpose of the test, because it wouldn't be testing the original code path.

mburdette replied:
hmmm, I feel like `localFileToAssetMetadata`  is more of a helper function for each test to use and could be altered. The expected attributes would still be populated because you would still be doing that logic inside of `FileToAssetMetadata`

mpobrien replied:
Additionally - we need to distinguish between "asset description existed in the file, but had an empty attributes list" vs. "asset description was not present in the file". Without having `hasAssetDesc` that wouldn't be possible.

mpobrien commented:
@mburdette updated - changed the function signature like we discussed, and refactored the test to more directly test ListLocalAssetMetadata without relying on external code paths. Also added some assertions for the cache data and a couple of drive-bys.

mburdette commented on hosting/hosting.go:
```diff
@@ -73,7 +94,7 @@ func FileToAssetMetadata(appID, path, assetPath string, info os.FileInfo, desc A
 	})
 
-	return NewAssetMetadata(appID, assetPath, generated, info.Size(), desc.Attrs), nil
+	return NewAssetMetadata(appID, assetPath, generated, info.Size(), attrs, info.ModTime().Unix()), nil
```
not sure we want to take the `ModTime` from the file here, `NewAssetMetadat` just uses the current time because it is supposed to reflect the time for the asset on s3 which technically should be when this file is imported?

mpobrien replied:
(Discussed offline)
